src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ThermalCompositionalMultiphaseFVM.hpp

```diff
@@ -0,0 +1,144 @@
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file ThermalCompositionalMultiphaseFVM.hpp
+ */
+
+#ifndef GEOSX_LINEARALGEBRA_INTERFACES_HYPREMGRTHERMALCOMPOSITIONALMULTIPHASEFVM_HPP_
+#define GEOSX_LINEARALGEBRA_INTERFACES_HYPREMGRTHERMALCOMPOSITIONALMULTIPHASEFVM_HPP_
+
+#include "linearAlgebra/interfaces/hypre/HypreMGR.hpp"
+
+namespace geosx
+{
+
+namespace hypre
+{
+
+namespace mgr
+{
+
+/**
+ * @brief ThermalCompositionalMultiphaseFVM strategy.
+ *
+ * Labels description stored in point_marker_array
+ *               0 = pressure
+ *               1 = density
+ *             ... = densities
+ *   numLabels - 2 = last density
+ *   numLabels - 1 = temperature
+ *
+ * 2-level MGR reduction strategy
+ *   - 1st level: eliminate the reservoir density associated with the volume constraint
+ *   - 2nd level: eliminate the other reservoir densities
+ *   - The coarse grid (pressure and temperature system) is solved with BoomerAMG with numFunctions==2.
+ *
+ */
+class ThermalCompositionalMultiphaseFVM : public MGRStrategyBase< 2 >
+{
+public:
+  /**
+   * @brief Constructor.
+   * @param numComponentsPerField array with number of components for each field
+   */
+  explicit ThermalCompositionalMultiphaseFVM( arrayView1d< int const > const & numComponentsPerField )
+    : MGRStrategyBase( LvArray::integerConversion< HYPRE_Int >( numComponentsPerField[0] ) )
+  {
+    // Level 0: eliminate last density which corresponds to the volume constraint equation
+    m_labels[0].resize( m_numBlocks - 2 );
+    std::iota( m_labels[0].begin(), m_labels[0].end(), 0 );
+    m_labels[0].push_back( m_numBlocks-1 ); // keep temperature
+    // Level 1: eliminate the other densities
+    m_labels[1].push_back( 0 ); // keep pressure
+    m_labels[1].push_back( m_numBlocks-1 ); // keep temperature
+
+    setupLabels();
+
+    m_levelFRelaxMethod[0]     = MGRFRelaxationMethod::singleLevel; //default, i.e. Jacobi
+    m_levelInterpType[0]       = MGRInterpolationType::jacobi; // Diagonal scaling (Jacobi)
+    m_levelRestrictType[0]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[0] = MGRCoarseGridMethod::galerkin; // Standard Galerkin
+
+    m_levelFRelaxMethod[1]     = MGRFRelaxationMethod::singleLevel; //default, i.e. Jacobi
+    m_levelInterpType[1]       = MGRInterpolationType::injection; // Injection
+    m_levelRestrictType[1]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[1] = MGRCoarseGridMethod::cprLikeBlockDiag; // Non-Galerkin Quasi-IMPES CPR
+
+    // Global smoother at each level, only do block-GS for the condensed system
+    m_levelSmoothType[1] = 1;
+    m_levelSmoothIters[1] = 1;
```

Comment:
I probably would rephrase the comment. I think the sense is that global smoothing is performed only on the first level, after reservoir density associated with the volume constraint have been eliminated. Although it is the default option (no global smoothing), we may add for clarity:
```c++
m_levelSmoothIters[0] = 0;
```

Not in this PR, but we should add enumerations to account for new capabilities introduced in #1712 after updating _hypre_.

Reply:
Yes, I added the `m_levelSmoothIters[0] = 0;` to make it clear that there is no global smoothing in this case. And yes, I agree that we should update the enumerations after the hypre update. I started adding some of them in `HypreUtils.hpp` in the last commit of #1712, but this should probably be checked again / made more uniform.